[Concept Entry] SwiftUI Tables (#3800)
Codecademy/docs: content/swiftui/concepts/tables/tables.md

---
Title: 'Tables' 
Description: 'A container that displays data in rows organized into one or more columns and optionally allows the selection of one or more items.'
Subjects: 
  - 'iOS'
  - 'Mobile Development'
Tags:
  - 'iOS'
  - 'SwiftUI'
  - 'Tables'
CatalogContent:
  - 'learn-ios-course'
  - 'paths/ios-developer'
---

> **`Table`** in SwiftUI is a container that enables the creation and presentation of data collections in a tabular format, organized into columns and rows. It is a valuable tool for displaying datasets that have a structured, grid-like layout. When the number of rows exceeds the available vertical space, the `Table` component will automatically enable vertical scrolling to ensure all data is accessible. Moreover, on macOS, horizontal scrolling becomes available when the number of columns exceeds the width of the view.

Reviewer: ```suggestion
A **`Table`** in SwiftUI is a container that enables the creation and presentation of data collections in a tabular format: values organized into columns and rows. It is a valuable tool for displaying datasets that have a structured, grid-like layout. When the number of rows exceeds the available vertical space, the `Table` component will automatically enable vertical scrolling to ensure all data is accessible. Moreover, on macOS, horizontal scrolling becomes available when the number of columns exceeds the width of the view.
```

> 

Reviewer: Before any examples there should be a syntax section to demonstrate how the structure is implemented generically with pseudo code. For example - 
https://www.codecademy.com/resources/docs/swiftui/views/button

## Example

`Table` is an excellent choice for presenting data in a concise, organized manner. In the example below, to create a straightforward three-column table from an array of `Singer`, the data model needs to conform to `Identifiable`.

```swift
struct ContentView: View {
    @State private var people = [
        Singer(name: "Taylor", lastName: "Swift", ranking: 95),
        Singer(name: "Olivia", lastName: "Rodrigo", ranking: 80),
        Singer(name: "Billie", lastName: "Eilish", ranking: 70),
        Singer(name: "Katy", lastName: "Perry", ranking: 75)
    ]
    
    var body: some View {
        Table(people) {
            TableColumn("Name", value: \.name)
            TableColumn("Last Name", value: \.lastName)
            TableColumn("Ranking") {
                people in
                Text(String(people.ranking))
            }
        }
    }
}

struct Singer: Identifiable {
    let name: String
    let lastName: String
    let ranking: Int
    let id = UUID()
    
    var fullName: String { name + " " + lastName}
}
```

> ### Adding sorting to a table

Reviewer: ```suggestion
### Adding Sorting to a Table
```

Here's an example of how to add sorting to the previous code:

```swift
struct ContentView: View {
  @State private var people = [
      Singer(name: "Taylor", lastName: "Swift", ranking: 95),
      Singer(name: "Olivia", lastName: "Rodrigo", ranking: 80),
      Singer(name: "Billie", lastName: "Eilish", ranking: 70),
      Singer(name: "Katy", lastName: "Perry", ranking: 75)
  ]
  
  @State private var sortOrder = [KeyPathComparator(\Singer.name)]
  
  var body: some View {
      Table(people, sortOrder: $sortOrder) {
          TableColumn("Name", value: \.name)
          TableColumn("Last Name", value: \.lastName)
          TableColumn("Ranking") {
              people in
              Text(String(people.ranking))
          }
      }
      .onChange(of: sortOrder) { newOrder in
          people.sort(using: newOrder)
      }
  }
}
```

> ### Adding selection to a table

Reviewer: ```suggestion
### Adding Selection to a Table
```

> struct ContentView: View {
>     @State private var people = [
>         Singer(name: "Taylor", lastName: "Swift", ranking: 95),
>         Singer(name: "Olivia", lastName: "Rodrigo", ranking: 80),
>         Singer(name: "Billie", lastName: "Eilish", ranking: 70),
>         Singer(name: "Katy", lastName: "Perry", ranking: 75)
>     ]
>     
>     @State private var sortOrder = [KeyPathComparator(\Singer.name)]
>     
>     var body: some View {
>         Table(people, sortOrder: $sortOrder) {
>             TableColumn("Name", value: \.name)
>             TableColumn("Last Name", value: \.lastName)
>             TableColumn("Ranking") {
>                 people in
>                 Text(String(people.ranking))
>             }
>         }
>         .onChange(of: sortOrder) { newOrder in
>             people.sort(using: newOrder)
>         }
>     }
> }

Reviewer 2: ```suggestion
struct ContentView: View {
  @State private var people = [
      Singer(name: "Taylor", lastName: "Swift", ranking: 95),
      Singer(name: "Olivia", lastName: "Rodrigo", ranking: 80),
      Singer(name: "Billie", lastName: "Eilish", ranking: 70),
      Singer(name: "Katy", lastName: "Perry", ranking: 75)
  ]
  
  @State private var sortOrder = [KeyPathComparator(\Singer.name)]
  
  var body: some View {
      Table(people, sortOrder: $sortOrder) {
          TableColumn("Name", value: \.name)
          TableColumn("Last Name", value: \.lastName)
          TableColumn("Ranking") {
              people in
              Text(String(people.ranking))
          }
      }
      .onChange(of: sortOrder) { newOrder in
          people.sort(using: newOrder)
      }
  }
}
```

To improve table interactivity, enabling users to select one or more rows can be achieved. To do this, binding to an optional identifier of the item should be provided to the `Table`. For creating a table that permits multiple selections, a binding to a Set of IDs can be established. Here's an example of how to incorporate multi-select functionality into the previous example:

```swift
 @State private var selection = Set<Singer.ID>()

var body: some View {
  Table(people, selection: $selection, sortOrder: $sortOrder) {...}
}
```

> ### Customizing a Table style

Reviewer: ```suggestion
### Customizing a Table Style
```

>  @State private var selection = Set<Singer.ID>()
> 
>  var body: some View {
>     Table(people, selection: $selection, sortOrder: $sortOrder) {...}
>  }

Reviewer 2: ```suggestion
 @State private var selection = Set<Singer.ID>()

var body: some View {
  Table(people, selection: $selection, sortOrder: $sortOrder) {...}
}
```

> One of the easiest methods to change a table's appearance is by using the `.tableStyle(_:)` modifier. This modifier requires one argument, a TableStyle, which defines the style to use for the table. For instance, SwiftUI provides several table styles, such as `InsetTableStyle` and, on macOS, `BorderedTableStyle`. The default style is `AutomaticTableStyle`, which is available on all platforms that support `Table`.

Reviewer: ```suggestion
One of the easiest methods to change a table's appearance is by using the `.tableStyle(_:)` modifier. This modifier requires just one argument, a predefined table style. SwiftUI provides several options, such as `InsetTableStyle` and, on macOS, `BorderedTableStyle`. The default style is `AutomaticTableStyle`, which is available on all platforms that support `Table`.
```

```swift
Table(people, selection: $selection, sortOrder: $sortOrder) {
} 
.tableStyle(InsertTableStyle())
```

> ### Setting a fixed width for the table column

Reviewer: ```suggestion
### Setting a Fixed Width for a Table Column
```

---
Title: 'Tables' 
Description: 'A container that displays data in rows organized into one or more columns and optionally allows the selection of one or more items.'
Subjects: 
  - 'iOS'
  - 'Mobile Development'
Tags:
  - 'iOS'
  - 'SwiftUI'
  - 'Tables'
CatalogContent:
  - 'learn-ios-course'
  - 'paths/ios-developer'
---

> 

Reviewer 2: ```suggestion
```
No need to give 2 white space here

> **Table** in SwiftUI is a structure that allows us to create and display collections of data in columns and rows. Many datasets have a tabular structure and can be efficiently presented in a table format. If the number of rows exceeds the available space, `Table` will enable vertical scrolling automatically. Additionally, on macOS, horizontal scrolling is available when the number of columns exceeds the view's width. 

Reviewer 2: ```suggestion
**`Table`** in SwiftUI is a container that enables the creation and presentation of data collections in a tabular format, organized into columns and rows. It is a valuable tool for displaying datasets that have a structured, grid-like layout. When the number of rows exceeds the available vertical space, the `Table` component will automatically enable vertical scrolling to ensure all data is accessible. Moreover, on macOS, horizontal scrolling becomes available when the number of columns exceeds the width of the view.
```

> ## Creating a Table 
> 
> `Table` is a great option for displaying a collection of data in a compact form. In the example below, in order to create a simple three-column table from an array of Singer your data model has to conform to `Identifiable`.
> 
> ### Example

Reviewer 2: ```suggestion
## Example

`Table` is an excellent choice for presenting data in a concise, organized manner. In the example below, to create a straightforward three-column table from an array of `Singer`, the data model needs to conform to `Identifiable`.

```
Let's remove the above heading and add it below the example heading only. Also refrain from using pronouns.

```swift
struct ContentView: View {
    @State private var people = [
        Singer(name: "Taylor", lastName: "Swift", ranking: 95),
        Singer(name: "Olivia", lastName: "Rodrigo", ranking: 80),
        Singer(name: "Billie", lastName: "Eilish", ranking: 70),
        Singer(name: "Katy", lastName: "Perry", ranking: 75)
    ]
    
    var body: some View {
        Table(people) {
            TableColumn("Name", value: \.name)
            TableColumn("Last Name", value: \.lastName)
            TableColumn("Ranking") {
                people in
                Text(String(people.ranking))
            }
        }
    }
}

struct Singer: Identifiable {
    let name: String
    let lastName: String
    let ranking: Int
    let id = UUID()
    
    var fullName: String { name + " " + lastName}
}
```

## Adding sorting to a table

> To make the columns of a `Table` sortable we need to make an array of KeyPathComparator objects with the default sorting inside.  Additionally, provide a binding to an array of sort comparators to watch for changes in the sort order. The table doesn't sort data by itself. To enable sorting, you can monitor changes by using the `.onChange(of:perform:)` modifier and then sort the data within the closure. 

Reviewer 2: ```suggestion
To enable sorting of `Table` columns, an array of `KeyPathComparator` objects should be created with default sorting inside. Additionally, a binding to an array of sort comparators needs to be provided to track changes in the sorting order. The table itself does not perform data sorting itself. To activate sorting, monitoring changes can be achieved using the `.onChange(of:perform:)` modifier and then sorting the data within the closure.
```

> Here's an example of how to add sorting to the previous code:

Reviewer 2: ```suggestion

Here's an example of how to add sorting to the previous code:
```

```swift
struct ContentView: View {
    @State private var people = [
        Singer(name: "Taylor", lastName: "Swift", ranking: 95),
        Singer(name: "Olivia", lastName: "Rodrigo", ranking: 80),
        Singer(name: "Billie", lastName: "Eilish", ranking: 70),
        Singer(name: "Katy", lastName: "Perry", ranking: 75)
    ]
    
    @State private var sortOrder = [KeyPathComparator(\Singer.name)]
    
    var body: some View {
        Table(people, sortOrder: $sortOrder) {
            TableColumn("Name", value: \.name)
            TableColumn("Last Name", value: \.lastName)
            TableColumn("Ranking") {
                people in
                Text(String(people.ranking))
            }
        }
        .onChange(of: sortOrder) { newOrder in
            people.sort(using: newOrder)
        }
    }
}
```

## Adding selection to a table

>  To enhance the interactivity of tables we can enable users to choose one or more rows. To achieve this pass the `Table` a binding to an optional identifier of the item. To create a table that allows for multiple selections, you can establish a binding to a Set of IDs. 
>  Here's an example of how to use multi-select functionality into the previous example:

Reviewer 2: ```suggestion
To improve table interactivity, enabling users to select one or more rows can be achieved. To do this, binding to an optional identifier of the item should be provided to the `Table`. For creating a table that permits multiple selections, a binding to a Set of IDs can be established. Here's an example of how to incorporate multi-select functionality into the previous example:
```

```swift

>  // @State private var selection: Singer.ID? allows one row selection

Reviewer 2: ```suggestion
```

 @State private var selection = Set<Singer.ID>()

 var body: some View {
    Table(people, selection: $selection, sortOrder: $sortOrder) {...}
 }
```

> ## Customizing a Table style

Reviewer 2: ```suggestion

## Customizing a Table style
```
Add a line before starting a new heading

One of the easiest methods to change a table's appearance is by using the `.tableStyle(_:)` modifier. This modifier requires one argument, a TableStyle, which defines the style to use for the table. For instance, SwiftUI provides several table styles, such as `InsetTableStyle` and, on macOS, `BorderedTableStyle`. The default style is `AutomaticTableStyle`, which is available on all platforms that support `Table`.

```swift
Table(people, selection: $selection, sortOrder: $sortOrder) {

>     // Table columns

Reviewer 2: ```suggestion
```

} 
.tableStyle(InsertTableStyle())
```

## Setting a fixed width for the table column

> You can set a specific width for a table column:

Reviewer 2: ```suggestion
Defining a precise width for a table column can be accomplished as follows:
```
